test/appstats_test.js

```diff
@@ -33,175 +35,456 @@ describe('App Stats tests', function() {
-                {'Updater':{'error':'Error: getaddrinfo ENOTFOUND test.table.core.windows.net test.table.core.windows.net:443'}}
-            ]
-        };
+        it('getStats inaccurate stats. Error in get message batch.', function(done) {
+            // Collection stats Azure mocks
+               nock('https://testappo365.queue.core.windows.net:443', {'encodedQueryParams':true})
+               .get('/alertlogic-stats')
+               .query({'comp':'metadata'})
+               .times(100)
+               .reply(200, '', mock.statsQueueMetadataHeaders(65));
+
+               nock('https://testappo365.queue.core.windows.net:443', {"encodedQueryParams":true})
+               .get('/alertlogic-stats/messages')
+               .query(true)
+               .times(2)
+               .reply(200, mock.statsMessages)
+               .get('/alertlogic-stats/messages')
+               .query(true)
+               .times(1)
+               .reply(403, mock.statsQueue403, mock.statsQueue403Headers);
+
+               nock('https://testappo365.queue.core.windows.net:443', {"encodedQueryParams":true})
+               .delete(/alertlogic-stats\/messages.*/)
+               .query(true)
+               .times(100)
+               .reply(204,'');
+
+               process.env.AzureWebJobsStorage = 'DefaultEndpointsProtocol=https;AccountName=testappo365;AccountKey=S0meKey+';
+               var collectionStats = new AzureCollectionStats(mock.DEFAULT_FUNCTION_CONTEXT);
+               
+               collectionStats.getStats(function(err, result) {
+                   const expected = { log: { bytes: 40, events: 60 } };
+                   assert.notEqual(err, null);
+                   assert.deepEqual(result, expected);
+                   done();
+               });
+           });
+
+        it('getStats inaccurate stats. Error in delete message.', function(done) {
+            // Collection stats Azure mocks
+               nock('https://testappo365.queue.core.windows.net:443', {'encodedQueryParams':true})
+               .get('/alertlogic-stats')
+               .query({'comp':'metadata'})
+               .times(100)
+               .reply(200, '', mock.statsQueueMetadataHeaders(65));
+
+               nock('https://testappo365.queue.core.windows.net:443', {"encodedQueryParams":true})
+               .get('/alertlogic-stats/messages')
+               .query(true)
+               .times(3)
+               .reply(200, mock.statsMessages);
+               
+               nock('https://testappo365.queue.core.windows.net:443', {"encodedQueryParams":true})
+               .delete(/alertlogic-stats\/messages.*/)
+               .query(true)
+               .times(2)
+               .reply(204,'')
+               .delete(/alertlogic-stats\/messages.*/)
+               .query(true)
+               .times(1)
+               .reply(403, mock.statsQueue403, mock.statsQueue403Headers);
```

Comment:
Here, if 2nd call returns an error, the 3rd call won't happen (IIUC) - is that ok?

Reply:
And, btw, if we get an error on deleteMessage, will that message (that's left in the queue, I guess) mess up further collection?

Reply:
I'll change `async.each` to `async.filter` and subtract undeleted stats from the aggregated value.